eras/alonzo/impl/src/Cardano/Ledger/Alonzo/TxWits.hs

```
@@ -579,55 +580,93 @@ instance
       txWitnessField 6 = fieldA addScriptsTxWitsRaw (decodeAlonzoPlutusScript SPlutusV2)
       txWitnessField 7 = fieldA addScriptsTxWitsRaw (decodeAlonzoPlutusScript SPlutusV3)
       txWitnessField n = invalidField n
       {-# INLINE txWitnessField #-}
 
-      nativeScriptsDecoder :: Decoder s (Annotator (Map ScriptHash (Script era)))
-      nativeScriptsDecoder =
-        ifDecoderVersionAtLeast
-          (natVersion @9)
-          ( allowTag setTag
-              >> mapTraverseableDecoderA (decodeNonEmptyList pairDecoder) (Map.fromList . NE.toList)
-          )
-          (mapTraverseableDecoderA (decodeList pairDecoder) Map.fromList)
+      pairDecoder :: Decoder s (Annotator (ScriptHash, Script era))
+      pairDecoder = fmap (asHashedScriptPair @era . fromNativeScript) <$> decCBOR
+      {-# INLINE pairDecoder #-}
+
+      noDuplicatesDatsDecoder :: Decoder s (Annotator (TxDats era))
+      noDuplicatesDatsDecoder = do
+        allowTag setTag
+        dats <- decodeList decCBOR
+        pure $ TxDats <$> go Map.empty dats
         where
-          pairDecoder :: Decoder s (Annotator (ScriptHash, Script era))
-          pairDecoder = fmap (asHashedScriptPair . fromNativeScript) <$> decCBOR
+          go m [] = pure m
+          go m (x : xs) = do
+            x' <- x
+            let dh = hashData x'
+            if dh `Map.member` m
+              then fail $ "Duplicate dats found: " <> show dh
+              else go (Map.insert dh x' m) xs
+      {-# INLINE noDuplicatesDatsDecoder #-}
+
+      noDuplicatesScriptsDecoder :: Decoder s (Annotator (Map ScriptHash (Script era)))
+      noDuplicatesScriptsDecoder = do
+        allowTag setTag
+        scripts <- decodeList $ fmap (fromNativeScript @era) <$> decCBOR
+        pure $ go Map.empty scripts
+        where
+          go m [] = pure m
+          go m (x : xs) = do
+            x' <- x
+            let sh = hashScript x'
+            if sh `Map.member` m
```

This approach of checking membership on every insert is slower when compared to keeping track of the length and then failing when number of elements in the Map/Set is smaller than in the original list. We don't need to report the exact duplicate, this can be left up to the user to figure out. So, this is how I'd suggest to write it:
```haskell
          go (m, n) [] = do
            when (Map.size == 0) $ fail "Empty script Set is not allowed"
            when (Map.size /= n) $ fail "Duplicate elements in the scripts Set were encountered"
            pure m
          go (!m, !n) (x : xs) =
            go (Map.insert (hashScript x) x m, n + 1) xs

---

My thinking was that this way we can fail early without having to decode the entire map and `O(n log n)` vs `O(n)` is not going to make a huge difference at the sizes we're dealing with. I can change it to your solution though if you think it makes more sense to do it that way.

---

But now that I think of it, scripts can be really small, so you could fit quite a bit of those into a transaction. In that case making the worst case as good as possible is a good idea.

---

> My thinking was that this way we can fail early without having to decode the entire map

We need to optimize our decoders for the positive case, when transaction is valid, because that the case that is being decoded the most. We don't need to worry ourselves too much about transactions that can't make it into the mempool due to them being malformed.

Naturally, we can't have dangerously slow decoders that fail decoding either, but that would be pretty hard to do anyways :grin: